src/Common/src/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs

```diff
@@ -393,6 +396,181 @@ private static ComputedInstanceFieldLayout ComputeSequentialFieldLayout(Metadata
+            //   - For non-value class fields, save the largest alignment we've seen
+            foreach (var field in type.GetFields())
+            {
+                if (field.IsStatic)
+                    continue;
+
+                if (IsByValueClass(field.FieldType))
+                {
+                    instanceValueClassFieldDescs.Add(field);
+                    continue;
+                }
+
+                var fieldSizeAndAlignment = ComputeFieldSizeAndAlignment(field.FieldType, packingSize);
+                int fieldSizeAsInt = fieldSizeAndAlignment.Size.AsInt;
+
+                maxInstanceFieldSize = Math.Max(fieldSizeAsInt, maxInstanceFieldSize);
+                largestAlignmentRequired = LayoutInt.Max(fieldSizeAndAlignment.Alignment, largestAlignmentRequired);
+
+                if (field.FieldType.IsGCPointer)
+                {
+                    if (!instanceGCPointerFieldsBySize.TryGetValue(fieldSizeAsInt, out List<FieldDesc> instanceFieldsWithMatchingSize))
+                    {
+                        instanceFieldsWithMatchingSize = new List<FieldDesc>();
+                        instanceGCPointerFieldsBySize[fieldSizeAsInt] = instanceFieldsWithMatchingSize;
+                    }
+                    instanceFieldsWithMatchingSize.Add(field);
+                }
+                else
+                {
+                    if (!instanceNonGCPointerFieldsBySize.TryGetValue(fieldSizeAsInt, out List<FieldDesc> instanceFieldsWithMatchingSize))
+                    {
+                        instanceFieldsWithMatchingSize = new List<FieldDesc>();
+                        instanceNonGCPointerFieldsBySize[fieldSizeAsInt] = instanceFieldsWithMatchingSize;
+                    }
+                    instanceFieldsWithMatchingSize.Add(field);
+                }
+            }
+
+            // Align the field offset to the required alignment of an int: 4
+            // Not sure why that is, but a .NET Core 2.1 app seems to align the subclass's fields automatically to this boundary
+            cumulativeInstanceFieldPos = LayoutInt.AlignUp(cumulativeInstanceFieldPos, new LayoutInt(4));
+
+            // First, place small fields immediately after the parent field bytes if there are a number of field bytes that are not aligned
+            // GC pointer fields and value class fields are not considered for this optimization
+            // Place the fields in order, starting with the largest size.
+            int parentByteOffsetModulo = cumulativeInstanceFieldPos.AsInt % type.Context.Target.PointerSize;
+            if (parentByteOffsetModulo != 0)
+            {
+                int alignmentBytesRemaining = type.BaseType.InstanceByteCount.AsInt - cumulativeInstanceFieldPos.AsInt;
+                bool continueProcessingInstanceFields = true;
+
+                for (int i = maxInstanceFieldSize; continueProcessingInstanceFields && i > 0; i /= 2)
```

Comment:
I assume this is some legacy CoreCLR behavior we must adhere to - otherwise the "i /= 2" doesn't make sense to me - I would assume this will end up skipping quite a few candidates.

Reply:
We assume the size of all fields is a power of 2. Unlike CoreCLR which tends to operate on indexes that really mean "log2(size)", this is just "size". This is why I'm dividing/multiplying by two when I change the index.

Reply:
I see, thanks for the explanation.